# North American Retailer Analysis

In [ ]:
import pandas as pd

#list of missing value types to be recognised
missing_val = ['?','#NAME?']


#Reading and converting csv data in to data frame
df_retail_sales=pd.read_csv('Project 2 - Retail Sales Dataset.csv',skipinitialspace = True, na_values = missing_val)

In [6]:
#Column names
df_retail_sales.columns

Index(['ORDER_NUMBER', 'SHIPMENT_NUMBER', 'ITEM_NUMBER', 'INVOICE_NUMBER',
       'INVOICE_DATE', 'ORDER_DATE', 'CUSTOMER_NUMBER', 'CUSTOMER_LOCATION',
       'CUSTOMER_DEPARTMENT', 'CUSTOMER_PROVINCE', 'ITEM_DESCRIPTION',
       'SHIPPING_REGION_IDENTIFIER', 'SHIPPING_REGION_CODE',
       'SHIPPING_REGION_NAME', 'SHIPPING_UNIT_MEASURE',
       'SHIPPING_PACK_QUANTITY', 'ITEM_UNIT_QUANTITY', 'ITEM_DEPARTMENT',
       'ITEM_SECTION', 'ITEM_SECTION_MODULE', 'ITEM_GROUP', 'ITEM_SEQUENCE',
       'ITEM_STOCK_STATUS', 'WAREHOUSE_IDENTIFIER', 'WAREHOUSE_NUMBER',
       'WAREHOUSE_NAME', 'WAREHOUSE_REGION_NAME', 'VENDOR_IDENTIFIER',
       'VENDOR_NUMBER', 'VENDOR_NAME', 'VENDOR_CITY', 'VENDOR_COUNTRY',
       'VENDOR_ADDRESS', 'ORDER_QUANTITY', 'SHIPPED_QUANTITY',
       'COMPANY_IDENTIFIER', 'COMPANY_CODE', 'COMPANY_NAME', 'COMPANY_NUMBER',
       'AVERAGE_COST_PRICE', 'SALES_AMOUNT', 'INVOICE_AMOUNT', 'COST_PRICE',
       'UPC_NUMBER', 'PURCHASE_ORDER_NUMBER', 'ORDER_TYPE'],
      dtype='o

In [7]:
#Checking the first 5 rows of the dataframe
df_retail_sales.head()

ORDER_NUMBER  SHIPMENT_NUMBER  ITEM_NUMBER INVOICE_NUMBER  INVOICE_DATE  \
0      6735217         20171129       428714        6735217      20171129   
1      6165708         20190626       408872        6165708      20190626   
2      6517073         20180910       354274        6517073      20180910   
3      6723693         20200513       222679        6723693      20200513   
4      6766514         20190128       431999        6766514      20190128   

   ORDER_DATE  CUSTOMER_NUMBER  CUSTOMER_LOCATION  CUSTOMER_DEPARTMENT  \
0    20171128              410                 10                    9   
1    20190624             1375                  1                   83   
2    20180908             2196                  1                   70   
3    20200509             1934                 50                   70   
4    20190125            75379                201                    0   

  CUSTOMER_PROVINCE  ... COMPANY_CODE       COMPANY_NAME COMPANY_NUMBER  \
0             TEXAS  ...        88880  BRANDS GLOBAL CO.              0   
1           ARIZONA  ...        88880  BRANDS GLOBAL CO.              0   
2        CALIFORNIA  ...        88880  BRANDS GLOBAL CO.              0   
3           ARIZONA  ...        88880  BRANDS GLOBAL CO.              0   
4             TEXAS  ...        88880  BRANDS GLOBAL CO.              0   

  AVERAGE_COST_PRICE SALES_AMOUNT  INVOICE_AMOUNT  COST_PRICE    UPC_NUMBER  \
0                0.0         3.36            0.56        0.56  6.224543e+10   
1                0.0        12.00            4.00        4.00  6.224543e+10   
2                0.0         7.16            1.79        1.79  6.224543e+10   
3                0.0        19.60            4.90        4.90  6.224543e+10   
4                0.0        62.58           10.43       10.43  6.224543e+10   

   PURCHASE_ORDER_NUMBER  ORDER_TYPE  
0                    0.0     REGULAR  
1                    0.0     REGULAR  
2                    0.0     REGULAR  
3                    0.0     REGULAR  
4                    0.0     REGULAR  

[5 rows x 46 columns]

In [8]:
#Shape of the dataframe
df_retail_sales.shape

(10000, 46)

In [9]:
#Summary of the dataframe
df_retail_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 46 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ORDER_NUMBER                10000 non-null  object 
 1   SHIPMENT_NUMBER             10000 non-null  int64  
 2   ITEM_NUMBER                 10000 non-null  int64  
 3   INVOICE_NUMBER              10000 non-null  object 
 4   INVOICE_DATE                10000 non-null  int64  
 5   ORDER_DATE                  10000 non-null  int64  
 6   CUSTOMER_NUMBER             10000 non-null  int64  
 7   CUSTOMER_LOCATION           10000 non-null  int64  
 8   CUSTOMER_DEPARTMENT         10000 non-null  int64  
 9   CUSTOMER_PROVINCE           9948 non-null   object 
 10  ITEM_DESCRIPTION            9928 non-null   object 
 11  SHIPPING_REGION_IDENTIFIER  10000 non-null  int64  
 12  SHIPPING_REGION_CODE        10000 non-null  object 
 13  SHIPPING_REGION_NAME        1000

By checking the data types for all the columns, it was noticed that the Invoice Date and Order Date columns were integers and hence not in a proper date-time format. 

In [10]:
#Converting date columns to date time format
df_retail_sales['INVOICE_DATE']= pd.to_datetime(df_retail_sales['INVOICE_DATE'],format='%Y%m%d')
df_retail_sales['ORDER_DATE']= pd.to_datetime(df_retail_sales['ORDER_DATE'],format='%Y%m%d')

In [11]:
#Checking if the date columns has changed to date time format correctly
df_retail_sales

ORDER_NUMBER  SHIPMENT_NUMBER  ITEM_NUMBER INVOICE_NUMBER INVOICE_DATE  \
0         6735217         20171129       428714        6735217   2017-11-29   
1         6165708         20190626       408872        6165708   2019-06-26   
2         6517073         20180910       354274        6517073   2018-09-10   
3         6723693         20200513       222679        6723693   2020-05-13   
4         6766514         20190128       431999        6766514   2019-01-28   
...           ...              ...          ...            ...          ...   
9995      5100802         20190403       522110        5100802   2019-04-03   
9996      5462559         20200827      4745048        5462559   2020-08-27   
9997      5165659         20190708       620658        5165659   2019-07-08   
9998      3274935         20200801       177550        3274935   2020-08-01   
9999      6104799         20201129       398867        6104799   2020-11-29   

     ORDER_DATE  CUSTOMER_NUMBER  CUSTOMER_LOCATION  CUSTOMER_DEPARTMENT  \
0    2017-11-28              410                 10                    9   
1    2019-06-24             1375                  1                   83   
2    2018-09-08             2196                  1                   70   
3    2020-05-09             1934                 50                   70   
4    2019-01-25            75379                201                    0   
...         ...              ...                ...                  ...   
9995 2019-03-29             1560                  4                   30   
9996 2020-08-25            75161                  1                    0   
9997 2019-07-07             1954                  1                   62   
9998 2020-07-31             2486                  4                   66   
9999 2020-11-23            15165                  1                    0   

     CUSTOMER_PROVINCE  ... COMPANY_CODE       COMPANY_NAME COMPANY_NUMBER  \
0                TEXAS  ...        88880  BRANDS GLOBAL CO.              0   
1              ARIZONA  ...        88880  BRANDS GLOBAL CO.              0   
2           CALIFORNIA  ...        88880  BRANDS GLOBAL CO.              0   
3              ARIZONA  ...        88880  BRANDS GLOBAL CO.              0   
4                TEXAS  ...        88880  BRANDS GLOBAL CO.              0   
...                ...  ...          ...                ...            ...   
9995          MICHIGAN  ...        88880  BRANDS GLOBAL CO.            513   
9996          MICHIGAN  ...        88880  BRANDS GLOBAL CO.            534   
9997          MICHIGAN  ...        88880  BRANDS GLOBAL CO.            528   
9998             TEXAS  ...        88880  BRANDS GLOBAL CO.              0   
9999        CALIFORNIA  ...        72223  THE GROCERY GROUP              0   

     AVERAGE_COST_PRICE SALES_AMOUNT  INVOICE_AMOUNT  COST_PRICE  \
0                   0.0         3.36            0.56        0.56   
1                   0.0        12.00            4.00        4.00   
2                   0.0         7.16            1.79        1.79   
3                   0.0        19.60            4.90        4.90   
4                   0.0        62.58           10.43       10.43   
...                 ...          ...             ...         ...   
9995                0.0         0.00            0.00        0.00   
9996                0.0         0.00            0.00        0.00   
9997                0.0         0.00            0.00        0.00   
9998                0.0         0.00            0.00        0.00   
9999                0.0         0.00            0.00        0.00   

        UPC_NUMBER  PURCHASE_ORDER_NUMBER  ORDER_TYPE  
0     6.224543e+10                    0.0     REGULAR  
1     6.224543e+10                    0.0     REGULAR  
2     6.224543e+10                    0.0     REGULAR  
3     6.224543e+10                    0.0     REGULAR  
4     6.224543e+10                    0.0     REGULAR  
...            ...                    ...         ...

In [12]:
#Checking for any null Values
df_retail_sales.isnull().sum()

ORDER_NUMBER                     0
SHIPMENT_NUMBER                  0
ITEM_NUMBER                      0
INVOICE_NUMBER                   0
INVOICE_DATE                     0
ORDER_DATE                       0
CUSTOMER_NUMBER                  0
CUSTOMER_LOCATION                0
CUSTOMER_DEPARTMENT              0
CUSTOMER_PROVINCE               52
ITEM_DESCRIPTION                72
SHIPPING_REGION_IDENTIFIER       0
SHIPPING_REGION_CODE             0
SHIPPING_REGION_NAME             0
SHIPPING_UNIT_MEASURE            0
SHIPPING_PACK_QUANTITY           0
ITEM_UNIT_QUANTITY               0
ITEM_DEPARTMENT                  0
ITEM_SECTION                     0
ITEM_SECTION_MODULE              0
ITEM_GROUP                       0
ITEM_SEQUENCE                    0
ITEM_STOCK_STATUS               52
WAREHOUSE_IDENTIFIER             0
WAREHOUSE_NUMBER                 0
WAREHOUSE_NAME                   0
WAREHOUSE_REGION_NAME            0
VENDOR_IDENTIFIER                0
VENDOR_NUMBER       

In [13]:
#Removing null values in certain columns
df_new = df_retail_sales.dropna(subset = ["CUSTOMER_PROVINCE","ITEM_STOCK_STATUS"])
df_new

#Shape of the new dataframe after removing null values
df_new.shape

(9948, 46)

Removed 52 null values in both the columns ‘Customer Province’ and ‘Item Stock Status’ as it was only a very small amount. After removing these null values, the number of rows was reduced from 10,000 to 9,948 rows.

In [14]:
#Removing certain columns that are not relevant
df_new.drop(["INVOICE_NUMBER","ITEM_DESCRIPTION","SHIPPING_REGION_IDENTIFIER","SHIPPING_REGION_CODE","WAREHOUSE_IDENTIFIER","WAREHOUSE_NUMBER","COMPANY_IDENTIFIER","COMPANY_CODE","COMPANY_NUMBER","VENDOR_NUMBER","VENDOR_ADDRESS","PURCHASE_ORDER_NUMBER"],axis=1,inplace=True)

C:\Users\Thara\anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


‘Item Description’ had 72 null values and it was removed since it couldn’t be used for any visualizations as it was too long. The columns ‘Item Number’ and ‘Item Department’ were used instead. There were columns that almost gave the same information for example ‘Shipping Region Identifier’, ‘Shipping Region Code’, and ‘Shipping Region Name’ therefore, only one column out of the three similar columns were taken and the other two columns were removed. The same rationale was applied to warehouse-related columns and company-related columns as well. Since ‘Vendor Number’, ‘Vendor City’, ‘Vendor Country’, and ‘Vendor Address’ had 2095 null values, those columns were removed and only the ‘Vendor Identifier’ column which had no null values were kept. ‘Invoice Number’ and ‘Purchase Order Number’ column was removed as they were considered not relevant for the analysis. 

In [15]:
df_new.columns

Index(['ORDER_NUMBER', 'SHIPMENT_NUMBER', 'ITEM_NUMBER', 'INVOICE_DATE',
       'ORDER_DATE', 'CUSTOMER_NUMBER', 'CUSTOMER_LOCATION',
       'CUSTOMER_DEPARTMENT', 'CUSTOMER_PROVINCE', 'SHIPPING_REGION_NAME',
       'SHIPPING_UNIT_MEASURE', 'SHIPPING_PACK_QUANTITY', 'ITEM_UNIT_QUANTITY',
       'ITEM_DEPARTMENT', 'ITEM_SECTION', 'ITEM_SECTION_MODULE', 'ITEM_GROUP',
       'ITEM_SEQUENCE', 'ITEM_STOCK_STATUS', 'WAREHOUSE_NAME',
       'WAREHOUSE_REGION_NAME', 'VENDOR_IDENTIFIER', 'VENDOR_NAME',
       'VENDOR_CITY', 'VENDOR_COUNTRY', 'ORDER_QUANTITY', 'SHIPPED_QUANTITY',
       'COMPANY_NAME', 'AVERAGE_COST_PRICE', 'SALES_AMOUNT', 'INVOICE_AMOUNT',
       'COST_PRICE', 'UPC_NUMBER', 'ORDER_TYPE'],
      dtype='object')

In [16]:
#Shape of the new dataframe after removing columns
df_new.shape

(9948, 34)

In [17]:
#to see first few rows 
df_new.head()

ORDER_NUMBER  SHIPMENT_NUMBER  ITEM_NUMBER INVOICE_DATE ORDER_DATE  \
0      6735217         20171129       428714   2017-11-29 2017-11-28   
1      6165708         20190626       408872   2019-06-26 2019-06-24   
2      6517073         20180910       354274   2018-09-10 2018-09-08   
3      6723693         20200513       222679   2020-05-13 2020-05-09   
4      6766514         20190128       431999   2019-01-28 2019-01-25   

   CUSTOMER_NUMBER  CUSTOMER_LOCATION  CUSTOMER_DEPARTMENT CUSTOMER_PROVINCE  \
0              410                 10                    9             TEXAS   
1             1375                  1                   83           ARIZONA   
2             2196                  1                   70        CALIFORNIA   
3             1934                 50                   70           ARIZONA   
4            75379                201                    0             TEXAS   

  SHIPPING_REGION_NAME  ... VENDOR_COUNTRY  ORDER_QUANTITY  SHIPPED_QUANTITY  \
0              FLORIDA  ...             US               6                 6   
1              FLORIDA  ...             US               3                 3   
2              FLORIDA  ...             US               4                 4   
3              FLORIDA  ...             US               4                 4   
4              FLORIDA  ...             US               6                 6   

        COMPANY_NAME  AVERAGE_COST_PRICE  SALES_AMOUNT  INVOICE_AMOUNT  \
0  BRANDS GLOBAL CO.                 0.0          3.36            0.56   
1  BRANDS GLOBAL CO.                 0.0         12.00            4.00   
2  BRANDS GLOBAL CO.                 0.0          7.16            1.79   
3  BRANDS GLOBAL CO.                 0.0         19.60            4.90   
4  BRANDS GLOBAL CO.                 0.0         62.58           10.43   

   COST_PRICE    UPC_NUMBER ORDER_TYPE  
0        0.56  6.224543e+10    REGULAR  
1        4.00  6.224543e+10    REGULAR  
2        1.79  6.224543e+10    REGULAR  
3        4.90  6.224543e+10    REGULAR  
4       10.43  6.224543e+10    REGULAR  

[5 rows x 34 columns]

#### Checking for any inconsistencies in Ware house name and Ware house Region Name

Inconsistencies between columns that should ideally be consistent with each other were checked such as the ‘Warehouse Name’ and ‘Warehouse Region Name’

In [18]:
warehouse_list = list(zip(df_new["WAREHOUSE_NAME"].values, df_new["WAREHOUSE_REGION_NAME"].values))
warehouse_list

[('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('FLORIDA HABS WAREHOUSE', 'FLORIDA'),
 ('OHIO FOOD WAREHOUSE', 'OHIO'),
 ('OHIO FOOD WAREHOUSE', 'OHIO'),
 ('TEXAS FOOD WAREHOUSE', 'TEXAS'),
 ('TEXAS FOOD WAREHOUSE', 'TEXAS'),
 ('OHIO FOOD WAREHOUSE', 'OHIO'),
 ('MISSOURI FOOD WAREHOUSE', 'MISSOURI'),
 ('OHIO FOOD WAREHOUSE', 'OHIO'),
 ('MISSOURI FOOD WAREHOUSE', 'MISSOURI'),
 ('FLORIDA FOOD WAREHOUSE', 'FLORIDA'),
 ('TEXAS FOOD WAREHOUSE', 'TEXAS'),
 ('MISSOURI FOOD WAREHOUSE', 'MISSOURI'),
 ('OHIO FOOD WAREHOUSE', 'OHIO'),
 ('MISSOURI FOOD WAREHOUSE', 'MISSOU

In [19]:
inconsistent = []
warehouse_dict = {}

for name, region in warehouse_list:
    
  if name not in warehouse_dict:
    warehouse_dict[name] = region
  else:
    if warehouse_dict[name] != region:
      inconsistent.append(name)

In [20]:
#Checking if there are any inconsistent records in the inconsistent list
inconsistent

[]


#### Checking for any inconsistencies in Item Number and Item Department

In [21]:
item_list = list(zip(df_new["ITEM_NUMBER"].values, df_new["ITEM_DEPARTMENT"].values))
item_list

[(428714, 40),
 (408872, 40),
 (354274, 40),
 (222679, 40),
 (431999, 40),
 (431155, 40),
 (313825, 40),
 (408773, 40),
 (429696, 40),
 (463927, 40),
 (428052, 40),
 (431973, 40),
 (5229190, 40),
 (592519, 30),
 (127894, 30),
 (2473791, 30),
 (171959, 30),
 (6265748, 30),
 (703744, 30),
 (947473, 30),
 (947473, 30),
 (2473791, 30),
 (911156, 30),
 (573410, 30),
 (841296, 30),
 (947465, 30),
 (680264, 30),
 (496877, 30),
 (496877, 30),
 (467936, 30),
 (668954, 30),
 (2415453, 30),
 (2448751, 30),
 (414110, 30),
 (171959, 30),
 (2415453, 30),
 (755017, 30),
 (6291926, 30),
 (503789, 30),
 (2472611, 30),
 (911156, 30),
 (733550, 30),
 (414110, 30),
 (947465, 30),
 (503789, 30),
 (735993, 30),
 (947473, 30),
 (921742, 30),
 (947465, 30),
 (703744, 30),
 (841296, 30),
 (275347, 30),
 (2478758, 30),
 (274423, 30),
 (593400, 30),
 (962829, 30),
 (2483493, 30),
 (592881, 30),
 (274423, 30),
 (632620, 30),
 (298448, 30),
 (467258, 30),
 (274423, 30),
 (2489425, 30),
 (219311, 30),
 (962779, 30)

In [22]:
inconsistent = []
item_dict = {}

for item, dep in item_list:
    
  if item not in item_dict:
    item_dict[item] = dep
  else:
    if item_dict[item]  != dep:
      inconsistent.append(item)

In [23]:
#Checking if there are any inconsistent records in the inconsistent list
inconsistent

[]

#### Checking for any inaccurate values for columns

Columns that shouldn’t ideally have negative values such as columns related to price and quantity were checked if there were any negative values. The results show that there were no negative values for the columns ‘Cost Price’, ‘Average Cost Price’, ‘Shipping Pack Quantity’, ‘Item Unit Quantity’, ‘Order Quantity, ‘Shipped Quantity’and ‘Sales Amount’.

In [24]:
df_new["ITEM_UNIT_QUANTITY"].min(), df_new["ITEM_UNIT_QUANTITY"].max()

(1.0, 2000.0)

In [25]:
df_new["INVOICE_AMOUNT"].min(), df_new["INVOICE_AMOUNT"].max()

(0.0, 20148.4)

In [26]:
df_new["SALES_AMOUNT"].min(), df_new["SALES_AMOUNT"].max()

(0.0, 25278.68)

In [27]:
df_new["COST_PRICE"].min(), df_new["COST_PRICE"].max()

(0.0, 20148.4)

In [28]:
df_new["AVERAGE_COST_PRICE"].min(), df_new["AVERAGE_COST_PRICE"].max()

(0.0, 424.16)

#### Creating a new column for Revenue, Cost and Profit

New columns for Revenue, Cost and Profit were created. Revenue was considered as ‘Sales Amount’ multiplied by the ‘Order Quantity’ and Cost was calculated as ‘Cost Price’ multiplied by the ‘Order Quantity’. Thereafter, Profit was calculated by deducting the ‘Cost’ from the ‘Revenue’

In [29]:
df_new["REVENUE"] = df_new["SALES_AMOUNT"] * df_new["ORDER_QUANTITY"]

C:\Users\Thara\AppData\Local\Temp/ipykernel_18780/759782189.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_new["REVENUE"] = df_new["SALES_AMOUNT"] * df_new["ORDER_QUANTITY"]


In [31]:
df_new["COST"] = df_new["COST_PRICE"] * df_new["ORDER_QUANTITY"]

C:\Users\Thara\AppData\Local\Temp/ipykernel_18780/3416835204.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_new["COST"] = df_new["COST_PRICE"] * df_new["ORDER_QUANTITY"]


In [32]:
df_new["PROFIT"] = df_new["REVENUE"] - df_new["COST"]

C:\Users\Thara\AppData\Local\Temp/ipykernel_18780/3960797594.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_new["PROFIT"] = df_new["REVENUE"] - df_new["COST"]


#### Exporting the cleaned data to a csv file

In [28]:
df_new.to_csv('Clean Retail Sales Dataset.csv',index=False)